arcgis-ios-sdk-samples/Layers/Export tiles/ExportTilesViewController.swift

```diff
@@ -17,154 +17,218 @@ import UIKit
         })
     }
     
-    @IBAction func closeAction() {
-        self.visualEffectView.isHidden = true
+    /// Get the extent within the extent view for generating a tile package.
+    func frameToExtent() -> AGSEnvelope {
+        let frame = mapView.convert(extentView.frame, from: self.view)
         
-        //release the map in order to free the tiled layer
-        self.previewMapView.map = nil
+        let minPoint = mapView.screen(toLocation: CGPoint(x: frame.minX, y: frame.minY))
+        let maxPoint = mapView.screen(toLocation: CGPoint(x: frame.maxX, y: frame.maxY))
+        let extent = AGSEnvelope(min: minPoint, max: maxPoint)
+        return extent
     }
     
-    private func deleteAllTpks() {
-        let documentDirectoryURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
-        do {
-            let files = try FileManager.default.contentsOfDirectory(atPath: documentDirectoryURL.path)
-            for file in files {
-                if file.hasSuffix(".tpk") {
-                    let url = documentDirectoryURL.appendingPathComponent(file)
-                    try FileManager.default.removeItem(at: url)
-                }
+    /// Make the destination URL for the tile package.
+    private func makeDownloadURL(fileFormat: TilePackageFormat) -> URL {
+        // If the downloadFileURL ends with ".tpk", the tile cache will use
+        // the legacy compact format. If the downloadFileURL ends with ".tpkx",
+        // the tile cache will use the current compact version 2 format.
+        // See more in the doc of
+        // `AGSExportTileCacheTask.exportTileCacheJob(with:downloadFileURL:)`.
+        
+        // Create the temp directory if it doesn't exist.
+        try? FileManager.default.createDirectory(at: temporaryDirectoryURL, withIntermediateDirectories: true)
+        return temporaryDirectoryURL
+            .appendingPathComponent("myTileCache", isDirectory: false)
+            .appendingPathExtension(fileFormat.fileExtension)
+    }
+    
+    // MARK: Actions
+    
+    @IBAction func exportTilesBarButtonTapped(_ sender: UIBarButtonItem) {
+        if let exportJob = job {
+            // If an existing job is downloading, cancel the download.
+            exportJob.progress.cancel()
+            job = nil
+        } else if let mapServiceInfo = exportTask.mapServiceInfo, mapServiceInfo.exportTilesAllowed {
+            // Otherwise, try to download when exporting tiles is allowed.
+            if mapServiceInfo.exportTileCacheCompactV2Allowed {
+                // Export using the CompactV2 (.tpkx) if it is supported.
+                self.initiateDownload(exportTask: self.exportTask, downloadFileURL: self.makeDownloadURL(fileFormat: .tpkx))
+            } else {
+                // Otherwise, use the CompactV1 (.tpk) format.
+                self.initiateDownload(exportTask: self.exportTask, downloadFileURL: self.makeDownloadURL(fileFormat: .tpk))
             }
```

Comment:
```suggestion
            let tilePackageFormat: TilePackageFormat
            if mapServiceInfo.exportTileCacheCompactV2Allowed {
                // Export using the CompactV2 (.tpkx) if it is supported.
                tilePackageFormat = .tpkx
            } else {
                // Otherwise, use the CompactV1 (.tpk) format.
                tilePackageFormat = .tpk
            }
            self.initiateDownload(exportTask: exportTask, downloadFileURL: makeDownloadURL(fileFormat: tilePackageFormat))
```